src/cgns_internals.c

```diff
@@ -187,73 +394,186 @@ int cgi_read_base(cgns_base *base)
         }
     }
 
-     /* Family_t */ /* -- FAMILY TREE -- */
-    if (cgi_get_nodes(base->id, "Family_t", &base->nfamilies, &id)) return CG_ERROR;
-    if (base->nfamilies>0) {
-         /* read & save families */
+    if (cgi_read_all_base_children(base->id, &nchildren, &childlist)) return CG_ERROR;
+    /* we now have all ids with label and name */
+    /* store them by label */
+    for (n = 0; n < nchildren; n++) {
+        nchildbylabel[childlist[n].type]++;
+    }
+    for (n = 0; n < NofBaseLabel; n++) {
+        if (nchildbylabel == 0) continue;
+        childbylabel[n] = CGNS_NEW(_childnode_t, nchildbylabel[n]);
+        nchildbylabel[n] = 0;
+    }
+    for (n = 0; n < nchildren; n++) {
+        memcpy((void *)&(childbylabel[childlist[n].type][nchildbylabel[childlist[n].type]]),
+            ((void *)&childlist[n]), sizeof(_childnode_t));
+        nchildbylabel[childlist[n].type] ++;
+    }
+    CGNS_FREE(childlist);
+    /* now consume the stacked childlist for each label */
+
+    /* Family_t */ /* -- FAMILY TREE -- */
+    int start = 0;
+    base->nfamilies = nchildbylabel[LabelFamily_t];
+    if (base->nfamilies > 0) {
+        /* read & save families */
         base->family = CGNS_NEW(cgns_family, base->nfamilies);
         for (n=0; n<base->nfamilies; n++) {
-            base->family[n].id = id[n];
-            base->family[n].link = cgi_read_link(id[n]);
+            base->family[n].id = childbylabel[LabelFamily_t][n].id;
+            base->family[n].link = cgi_read_link(childbylabel[LabelFamily_t][n].id);
             base->family[n].in_link = 0;
-            if (cgi_read_family(&base->family[n])) return CG_ERROR;
+            if (cgi_read_family(&base->family[n])) {
+                for (m = 0; m < NofBaseLabel; m++) {
+                    if (childbylabel[m] == NULL) continue;
+                    CGNS_FREE(childbylabel[m])
+                }
+                return CG_ERROR;
+            }
         }
-        CGNS_FREE(id);
     }
 
-     /* ReferenceState_t */
-    if (cgi_read_state(0, base->id, &base->state)) return CG_ERROR;
+    /* ReferenceState_t */
```

Comment:
It seems like maybe all of these could call a common function instead of repeating the body of the `if` statement for each...   Not that much code, but it is repeated eleven times...

Reply:
I have wrapped the "if" in a macro to make it cleaner.